File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day2/AcWing1249/main.go
package main
/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/28 15:19
 */
import (
    "bufio"
    "fmt"
    "os"
    "strconv"
)

//N const
const N int = 2e4 + 10
var p [N]int
var out = bufio.NewWriter(os.Stdout)
var sc = bufio.NewScanner(os.Stdin)
func main() {
    sc.Split(bufio.ScanWords)
    n, m := scanInt(), scanInt()
    for i := 1; i <= n; i++ {
        p[i] = i;
    }
    for ; m > 0; m-- {
        a, b := scanInt(), scanInt()
        a, b = find(a), find(b)
        if a != b {
            p[b] = a
        }
    }
    for q := scanInt(); q > 0; q-- {
        c, d := scanInt(), scanInt()
        if find(c) == find(d) {
            fmt.Fprintln(out, "Yes")
        } else {
            fmt.Fprintln(out, "No")
        }
    }
    fmt.Fprint(out)
    out.Flush()
}

func find(x int) int {
    for x != p[x] {
        p[x] = p[p[x]]
        x = p[x]
    }
    return x
}

func scanInt() int {
    sc.Scan()
    a, _ := strconv.Atoi(sc.Text())
    return a
}

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day2/AcWing1249/main.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/28 15:19
 */
using namespace std;

const int N = 20010;

int n, m;
int p[N];

int find(int x)
{
    if (p[x] != x) p[x] = find(p[x]);
    return p[x];
}

int main()
{
    scanf("%d%d", &n, &m);
    for (int i = 1; i <= n; i ++ ) p[i] = i;
    while (m -- )
    {
        int a, b;
        scanf("%d%d", &a, &b);
        p[find(a)] = find(b);
    }

    scanf("%d", &m);
    while (m -- )
    {
        int a, b;
        scanf("%d%d", &a, &b);
        if (find(a) == find(b)) puts("Yes");
        else puts("No");
    }

    return 0;
}

